activesupport/lib/active_support/core_ext/module/deprecation.rb

```diff
@@ -1,25 +1,27 @@
 # frozen_string_literal: true
 
 class Module
-  #   deprecate :foo
-  #   deprecate bar: 'message'
-  #   deprecate :foo, :bar, baz: 'warning!', qux: 'gone!'
+  #   deprecate :foo, deprecator: MyLib.deprecator
+  #   deprecate :foo, bar: "warning!", deprecator: MyLib.deprecator
   #
-  # You can also use custom deprecator instance:
-  #
-  #   deprecate :foo, deprecator: MyLib::Deprecator.new
-  #   deprecate :foo, bar: "warning!", deprecator: MyLib::Deprecator.new
-  #
-  # \Custom deprecators must respond to <tt>deprecation_warning(deprecated_method_name, message, caller_backtrace)</tt>
-  # method where you can implement your custom warning behavior.
+  # A deprecator is typically an instance of ActiveSupport::Deprecation, but you can also pass any object that responds
+  # to <tt>deprecation_warning(deprecated_method_name, message, caller_backtrace)</tt> where you can implement your
+  # custom warning behavior.
   #
   #   class MyLib::Deprecator
   #     def deprecation_warning(deprecated_method_name, message, caller_backtrace = nil)
   #       message = "#{deprecated_method_name} is deprecated and will be removed from MyLibrary | #{message}"
   #       Kernel.warn message
   #     end
   #   end
-  def deprecate(*method_names)
-    ActiveSupport::Deprecation.deprecate_methods(self, *method_names)
+  def deprecate(*method_names, deprecator: nil, **options)
+    if deprecator.is_a?(ActiveSupport::Deprecation)
+      deprecator.deprecate_methods(self, *method_names, **options)
+    elsif deprecator
+      ActiveSupport::Deprecation._instance.deprecate_methods(self, *method_names, **options, deprecator: deprecator)
```

Comment:
We will probably remove `ActiveSupport::Deprecation._instance` whenever we remove `ActiveSupport::Deprecation.instance`.  So I think eventually this will need to be:

```suggestion
      ActiveSupport.deprecator.deprecate_methods(self, *method_names, **options, deprecator: deprecator)
```

Reply:
> I think there's another discussion to have about this duck-typed API (only implementing `deprecation_warning`), it's not very Rails-y, and does not offer much value since you can just pass a real object, I'm not sure why this both added a way to instantiate deprecators and the duck-typed API: [71993c6](https://github.com/rails/rails/commit/71993c6f9770b1350aa41fe8c68f1dd2c7800403)

I agree, though I might be missing something.  It seems like the intent was to make the API more "lightweight" so that the user doesn't have to be aware of the `ActiveSupport::Deprecation` class.  But, like you said, the value of that is questionable.  Also, if a lightweight variant is desirable, a callable (e.g. a `Proc`) seems like a better choice.